[Smart Lists] Add support for parsing Smart List marker formats

rr-codes opened the pull request:
#### a03e91423aaffbc1ec753eccb1c7f973e6798177
<pre>
[Smart Lists] Add support for parsing Smart List marker formats
<a href="https://bugs.webkit.org/show_bug.cgi?id=298831">https://bugs.webkit.org/show_bug.cgi?id=298831</a>
<a href="https://rdar.apple.com/160544109">rdar://160544109</a>

Reviewed by Abrar Rahman Protyasha.

Add initial implementation support for Smart Lists, in particular the parser. From a high-level
perspective, this adds the following flow:

1. When an InsertTextCommand occurs, the command is intercepted by the Smart List logic if the
text is a space character (along with other conditions described in inline comments).

2. When this happens, the space character is not inputted. Instead, a TextList value is parsed
from the beginning of the current line up to the current position, which describes the list type,
start value, and style type.

3. The corresponding list element is then created and inserted, and the selection becomes inside of its list element child.

4. The appropriate style and attributes are set on the list element.

This is a simple implementation mainly to exercise the String -&gt; TextList parser (described below and in comments) and facilitate testing.

* Source/WTF/wtf/unicode/CharacterNames.h:

Add em-dash, used for dashed list markers.

* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::isSmartListsEnabled):
(WebCore::Editor::toggleSmartLists):

Add additionally guards to exit early if smart lists are unavailable by the system.

* Source/WebCore/editing/InsertTextCommand.cpp:
(WebCore::setListStyleTypeProperty):

Apply the ListStyleType value as an inline property on the specified element by converting the ListStyleType to a Ref&lt;CSSValue&gt;.

(WebCore::InsertTextCommand::applySmartListsIfNeeded):

If the criteria is met, discard the current command and subtitute an InsertListCommand in its place, using the parsed TextList. Then, apply the relevant attributes and styles, and delete the selection to remove the marker token characters.

(WebCore::InsertTextCommand::doApply):

Skip the rest of the application of this command if a Smart List was applied.

* Source/WebCore/editing/InsertTextCommand.h:

Add deeclaration of private helper method.

* Source/WebCore/editing/TextListParser.cpp: Added.
(WebCore::consumeNumber):
(WebCore::tryConsumeUnorderedDiscTextList):
(WebCore::tryConsumeUnorderedDashTextList):
(WebCore::tryConsumeOrderedDecimalTextList):
(WebCore::consumeTextList):
(WebCore::parseTextList):
* Source/WebCore/editing/TextListParser.h: Added.

Implement a rudimentary parser that converts a String to a TextList. The algorithm is described inline.

* Source/WebCore/style/values/lists/StyleListStyleType.cpp:
(WebCore::Style::ListStyleType::isDecimal const):

Add an accessor function to determine if the counter style type is Decimal, like some other of the existing functions.

(WebCore::Style::CSSValueCreation&lt;ListStyleType&gt;::operator):

It is currently possible to convert a CSSValue to a ListStyleType, however the opposite was not possible.
This adds support for that using the standard CSSValueCreation pattern for better composition, by creating a CSSValue depending on the value of the list style type.

* Source/WebCore/style/values/lists/StyleListStyleType.h:
(WebCore::Style::ListStyleType::ListStyleType):

Add a convenience constructor for the decimal style, like some other of the constructors.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/SmartLists.mm:
(webViewWithMarkup):
(insertText):
(InputInstruction::InputInstruction):
(keyDataForKeyInstruction):
(runTest):
(TEST(SmartLists, InsertingSpaceAndTextAfterBulletPointGeneratesListWithText)):
(TEST(SmartLists, InsertingSpaceAndTextAfterHyphenGeneratesDashedList)):
(TEST(SmartLists, InsertingSpaceAfterBulletPointGeneratesEmptyList)):
(TEST(SmartLists, InsertingSpaceAfterBulletPointInMiddleOfSentenceDoesNotGenerateList)):
(TEST(SmartLists, InsertingSpaceAfterPeriodAtStartOfSentenceDoesNotGenerateList)):
(TEST(SmartLists, InsertingSpaceAfterNumberGeneratesOrderedList)):
(TEST(SmartLists, InsertingSpaceAfterMultipleDigitNumberGeneratesOrderedList)):
(TEST(SmartLists, InsertingSpaceAfterInvalidNumberDoesNotGenerateOrderedList)):
(TEST(SmartLists, InsertingListMergesWithPreviousListIfPossible)):

Tests. Lots of tests!

Canonical link: <a href="https://commits.webkit.org/300101@main">https://commits.webkit.org/300101@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/791b08c5a1161f8696a340a7a1ad1e3731befdba

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121345 "23 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41042 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31701 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127784 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73426 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9f2a702a-b076-4683-b4b4-b7de25b5883a) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/123221 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41744 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49621 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92173 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/61331 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/fd9d80ae-f67a-40bb-9f8c-5b1aec7b212c) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124297 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33334 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72850 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/675771a9-98b4-4660-b3e8-0596e3dc2b41) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32351 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26869 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71364 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/113475 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102827 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27044 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130619 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/119865 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48273 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36701 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100770 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48641 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104929 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100675 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25515 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46087 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24150 "Passed tests") | [💥 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/44967 "An unexpected error occured. Recent messages:Printed configuration") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48131 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53844 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/150027 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47603 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/38154 "Found 1 new JSC stress test failure: stress/dont-link-virtual-calls-on-compiler-thread.js.no-llint (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50949 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49285 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

lilyspiniolas commented on Source/WTF/wtf/unicode/CharacterNames.h:
```diff
@@ -219,6 +220,7 @@ using WTF::Unicode::emSpace;
 using WTF::Unicode::emojiCat;
 using WTF::Unicode::emojiVariationSelector;
 using WTF::Unicode::enDash;
+using WTF::Unicode::emDash;
```
nit:
```suggestion
using WTF::Unicode::emDash;
using WTF::Unicode::enDash;
```

rr-codes replied:
LOL very good catch

lilyspiniolas commented on Source/WebCore/editing/InsertTextCommand.cpp:
```diff
@@ -131,6 +135,92 @@ bool InsertTextCommand::performOverwrite(const String& text, bool selectInserted
+        return;
+    }
+
+    auto& style = renderer->style();
```
```suggestion
    CheckedPtr style = renderer->style();
```

aprotyas commented on Source/WebCore/editing/TextListParser.cpp:
```diff
@@ -0,0 +1,154 @@
+constexpr int consumeNumber(StringParsingBuffer<Character>& input)
+{
+    // Parse the digits until there is no more input left or a non-ASCII digit character has been encountered.
+    WTF::Checked<int> value;
```
nit: style convention is to avoid the `WTF::` namespace prefix (given `using WTF::Checked;` exists)

aprotyas commented on Source/WebCore/editing/TextListParser.cpp:
```diff
@@ -0,0 +1,154 @@
+        value = (value * 10) + digitValue;
+    } while (!input.atEnd() && WTF::isASCIIDigit(*input));
+
+    ASSERT(value.value() > 0);
```
can it not be 0?

rr-codes replied:
I don’t believe so because of the earlier `*input == ‘0’` guard

rniwa commented on Source/WebCore/editing/InsertTextCommand.cpp:
```diff
@@ -131,6 +135,92 @@ bool InsertTextCommand::performOverwrite(const String& text, bool selectInserted
+
+        // This is either a newly created list, or an existing list that was just appended to.
+        // In the case of the latter, the existing list's ordering takes precedent over any new elements.
+        listElement->setAttributeWithoutOverwriting(HTMLNames::startAttr, start);
```
This will likely break undo/redo. Use `CompositeEditCommand::setNodeAttribute` instead.

rniwa commented on Source/WebCore/editing/InsertTextCommand.cpp:
```diff
@@ -131,6 +135,92 @@ bool InsertTextCommand::performOverwrite(const String& text, bool selectInserted
+    auto& pool = CSSValuePool::singleton();
+
+    Ref value = Style::createCSSValue(pool, style, styleType);
+    element.setInlineStyleProperty(CSSPropertyListStyleType, WTFMove(value));
```
Unfortunately we can't use `setInlineStyleProperty` directly like this. Use `setNodeAttribute` on `styleAttr` instead.

rr-codes replied:
Ah, ok. Out of curiosity, why can't we?

rniwa commented on Source/WebCore/editing/InsertTextCommand.cpp:
```diff
@@ -131,6 +135,92 @@ bool InsertTextCommand::performOverwrite(const String& text, bool selectInserted
+bool InsertTextCommand::applySmartListsIfNeeded()
+{
+    if (!document().editor().isSmartListsEnabled()) {
+        // Smart Lists are unavailable by the system or disabled by the user.
```
I don't think this comment is useful. Remove?

rniwa replied:
I don't think we should be adding these comments that simply repeat the code.

rr-codes replied:
ya, makes sense; I haven’t addressed all your comments yet, just trying to get it to build in EWS atm!

rniwa commented on Source/WebCore/editing/TextListParser.cpp:
```diff
@@ -0,0 +1,154 @@
+// MARK: Helpers
+
+template<typename Character>
+constexpr int consumeNumber(StringParsingBuffer<Character>& input)
```
Surely we can use `parseInteger` in `StringToIntegerConversion.h`?

rr-codes replied:
I wanted to use it, but I wasn’t able to for several reasons:

1. parseInteger accepts leading - or + signs, which I do not want
2. It has unnecessary extra added logic that I don’t need
3. It does not support StringParsingBuffer and will not consume it

Given these reasons, and the fact that only 4 lines of code are actually duplicated, I decided to do it this way.

rniwa commented on Source/WebCore/editing/TextListParser.cpp:
```diff
@@ -0,0 +1,154 @@
+        if (input.atEnd())
+            return { { { CSS::Keyword::Disc { } }, 0, false } };
+
+        // Invalid; skip to end.
```
Instead of adding a comment like this, can we have a helper function with a descriptive name: e.g.
`skipToEndForInvalidInput(input)`

rniwa replied:
This comment still withstands.

rr-codes replied:
ya, makes sense; I haven’t addressed all your comments yet, just trying to get it to build in EWS atm!

rniwa commented on Source/WebCore/editing/TextListParser.cpp:
```diff
@@ -0,0 +1,154 @@
+template<typename Character>
+inline std::optional<TextList> consumeTextList(StringParsingBuffer<Character>& input)
+{
+    // Consume U+002A or U+2022 if possible.
```
I don't think these comments are useful. Remove?

rniwa replied:
Again, I don't think we should be adding comments like this. It makes the code less readable.

rniwa commented on Source/WebCore/editing/TextListParser.cpp:
```diff
@@ -0,0 +1,154 @@
+    if (auto result = tryConsumeUnorderedDiscTextList(input))
+        return result;
+
+    // Consume U+2010 if possible.
```
Ditto.

rniwa replied:
Ditto.

rniwa commented on Source/WebCore/editing/TextListParser.cpp:
```diff
@@ -0,0 +1,154 @@
+    if (auto result = tryConsumeUnorderedDashTextList(input))
+        return result;
+
+    // Consume an ordinal followed by U+002E or U+0029 if possible.
```
Ditto.

rniwa commented on Source/WebCore/editing/TextListParser.cpp:
```diff
@@ -0,0 +1,154 @@
+template<typename Character>
+std::optional<TextList> tryConsumeOrderedDecimalTextList(StringParsingBuffer<Character>& input)
+{
+    // This algorithm is similar to the one in StringToIntegerConversion.h, but is stricter and simpler; specifically:
```
I don't think we want to duplicate this much code. I think we should do a simple string validation and use `parseInteger` to get the integer out of the string.

rniwa replied:
This comment still withstands.

rniwa commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SmartLists.mm:
```diff
@@ -262,4 +355,173 @@ static void setUserDefaultsValue(BOOL value)
+    } };
+
+    static constexpr std::initializer_list<InputInstruction> inputWithAsterisk {
+        '*', ' ', 'H', 'e', 'l', 'l', 'o'
```
This is hard to read. Can we instead use a string literal? e.g. `"* Hello"`.

rr-codes committed:
[Smart Lists] Add support for parsing Smart List marker formats
https://bugs.webkit.org/show_bug.cgi?id=298831
rdar://160544109

Reviewed by Abrar Rahman Protyasha.

Add initial implementation support for Smart Lists, in particular the parser. From a high-level
perspective, this adds the following flow:

1. When an InsertTextCommand occurs, the command is intercepted by the Smart List logic if the
text is a space character (along with other conditions described in inline comments).

2. When this happens, the space character is not inputted. Instead, a TextList value is parsed
from the beginning of the current line up to the current position, which describes the list type,
start value, and style type.

3. The corresponding list element is then created and inserted, and the selection becomes inside of its list element child.

4. The appropriate style and attributes are set on the list element.

This is a simple implementation mainly to exercise the String -> TextList parser (described below and in comments) and facilitate testing.

* Source/WTF/wtf/unicode/CharacterNames.h:

Add em-dash, used for dashed list markers.

* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::isSmartListsEnabled):
(WebCore::Editor::toggleSmartLists):

Add additionally guards to exit early if smart lists are unavailable by the system.

* Source/WebCore/editing/InsertTextCommand.cpp:
(WebCore::setListStyleTypeProperty):

Apply the ListStyleType value as an inline property on the specified element by converting the ListStyleType to a Ref<CSSValue>.

(WebCore::InsertTextCommand::applySmartListsIfNeeded):

If the criteria is met, discard the current command and subtitute an InsertListCommand in its place, using the parsed TextList. Then, apply the relevant attributes and styles, and delete the selection to remove the marker token characters.

(WebCore::InsertTextCommand::doApply):

Skip the rest of the application of this command if a Smart List was applied.

* Source/WebCore/editing/InsertTextCommand.h:

Add deeclaration of private helper method.

* Source/WebCore/editing/TextListParser.cpp: Added.
(WebCore::consumeNumber):
(WebCore::tryConsumeUnorderedDiscTextList):
(WebCore::tryConsumeUnorderedDashTextList):
(WebCore::tryConsumeOrderedDecimalTextList):
(WebCore::consumeTextList):
(WebCore::parseTextList):
* Source/WebCore/editing/TextListParser.h: Added.

Implement a rudimentary parser that converts a String to a TextList. The algorithm is described inline.

* Source/WebCore/style/values/lists/StyleListStyleType.cpp:
(WebCore::Style::ListStyleType::isDecimal const):

Add an accessor function to determine if the counter style type is Decimal, like some other of the existing functions.

(WebCore::Style::CSSValueCreation<ListStyleType>::operator):

It is currently possible to convert a CSSValue to a ListStyleType, however the opposite was not possible.
This adds support for that using the standard CSSValueCreation pattern for better composition, by creating a CSSValue depending on the value of the list style type.

* Source/WebCore/style/values/lists/StyleListStyleType.h:
(WebCore::Style::ListStyleType::ListStyleType):

Add a convenience constructor for the decimal style, like some other of the constructors.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/SmartLists.mm:
(webViewWithMarkup):
(insertText):
(InputInstruction::InputInstruction):
(keyDataForKeyInstruction):
(runTest):
(TEST(SmartLists, InsertingSpaceAndTextAfterBulletPointGeneratesListWithText)):
(TEST(SmartLists, InsertingSpaceAndTextAfterHyphenGeneratesDashedList)):
(TEST(SmartLists, InsertingSpaceAfterBulletPointGeneratesEmptyList)):
(TEST(SmartLists, InsertingSpaceAfterBulletPointInMiddleOfSentenceDoesNotGenerateList)):
(TEST(SmartLists, InsertingSpaceAfterPeriodAtStartOfSentenceDoesNotGenerateList)):
(TEST(SmartLists, InsertingSpaceAfterNumberGeneratesOrderedList)):
(TEST(SmartLists, InsertingSpaceAfterMultipleDigitNumberGeneratesOrderedList)):
(TEST(SmartLists, InsertingSpaceAfterInvalidNumberDoesNotGenerateOrderedList)):
(TEST(SmartLists, InsertingListMergesWithPreviousListIfPossible)):

Tests. Lots of tests!

Canonical link: https://commits.webkit.org/300101@main
